pkg/storage/formattedstore/batchwrite.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package formattedstore
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hyperledger/aries-framework-go/pkg/storage/edv/models"
+)
+
+// ErrValueIsDeleted is returned when value is deleted.
+var ErrValueIsDeleted = fmt.Errorf("value is deleted")
+
+type batchProvider interface {
+	Batch(batch *models.Batch) error
+}
+
+type batchStore interface {
+	AddEncryptedIndices(k string, encryptedDocumentBytes []byte) (*models.EncryptedDocument, error)
+}
+
+// BatchService is a batch service.
+type BatchService struct {
+	sizeLimit int
+	timeLimit time.Duration
+	sync.RWMutex
+	values       models.Batch
+	futureValues map[string]*batch
+	formatter    Formatter
+	provider     batchProvider // TODO remove batch provider after refactoring edv batch
+}
+
+type batch struct {
+	keyID     string
+	value     []byte
+	isDeleted bool
+}
+
+// NewBatchWrite new batch write.
+func NewBatchWrite(sizeLimit int, timeLimit time.Duration, formatter Formatter,
+	provider batchProvider) *BatchService {
+	ticker := time.NewTicker(timeLimit)
+	quit := make(chan struct{})
+	b := &BatchService{
+		sizeLimit: sizeLimit, timeLimit: timeLimit, values: make(models.Batch, 0),
+		futureValues: make(map[string]*batch), formatter: formatter, provider: provider,
+	}
+
+	go func(b *BatchService) {
+		for {
+			select {
+			case <-ticker.C:
+				if err := b.flush(); err != nil {
+					logger.Errorf(err.Error())
+				}
+			case <-quit:
+				ticker.Stop()
+				return
+			}
+		}
+	}(b)
+
+	return b
+}
+
+// Get value from batch values.
+func (b *BatchService) Get(k string) ([]byte, error) {
+	b.RLock()
+	v, ok := b.futureValues[k]
+	b.RUnlock()
+
+	if !ok {
+		return nil, fmt.Errorf("k %s not found", k)
+	}
+
+	if v.isDeleted {
+		return nil, ErrValueIsDeleted
+	}
+
+	return v.value, nil
+}
+
+// Put in batch value.
+func (b *BatchService) Put(s batchStore, k string, v []byte) error {
+	b.RLock()
+	size := len(b.values)
+	b.RUnlock()
+
+	if size >= b.sizeLimit {
+		if err := b.flush(); err != nil {
```

Comment:
Don't you need to put the new k/v before flushing?

Reply:
I will flush before adding new values